src/MongoDB.Driver.Core/Core/NativeLibraryLoader/LibraryLoader.cs

```diff
@@ -0,0 +1,134 @@
+        // public methods
+        public T GetDelegate<T>(string name)
+        {
+            IntPtr ptr = _nativeLoader.GetFunctionPointer(name);
+            if (ptr == IntPtr.Zero)
+            {
+                throw new TypeLoadException($"The function {name} was not found.");
+            }
+
+            return Marshal.GetDelegateForFunctionPointer<T>(ptr);
+        }
+
+        // private methods
+        private INativeLibraryLoader CreateNativeLoader(Func<SupportedPlatform, string> libraryLocator)
+        {
+            var currentPlatform = GetCurrentPlatform();
+            var relativePath = libraryLocator(currentPlatform);
+            var absolutePath = GetAbsolutePath(relativePath);
+            return CreateNativeLoader(currentPlatform, absolutePath);
+        }
+
+        private INativeLibraryLoader CreateNativeLoader(SupportedPlatform currentPlatform, string libraryPath)
+        {
+            switch (currentPlatform)
+            {
+                case SupportedPlatform.Linux:
+                    return new LinuxLibraryLoader(libraryPath);
+                case SupportedPlatform.MacOS:
+                    return new DarwinLibraryLoader(libraryPath);
+                case SupportedPlatform.Windows:
+                    return new WindowsLibraryLoader(libraryPath);
+                default:
+                    throw new PlatformNotSupportedException($"Unexpected platform {currentPlatform}.");
+            }
+        }
+
+        private string FindLibraryOrThrow(string basePath, string relativePath)
+        {
+            var fullPath = Path.Combine(basePath, relativePath);
+            if (File.Exists(fullPath))
+            {
+                return fullPath;
+            }
+
+            throw new FileNotFoundException($"Could not find library {fullPath}.");
+        }
+
+        private string GetAbsolutePath(string relativePath)
+        {
+            var assembly = typeof(LibraryLoader).GetTypeInfo().Assembly;
+            var codeBase = assembly.CodeBase;
+            var uri = new Uri(codeBase);
+            var location = uri.AbsolutePath;
+            var basepath = Path.GetDirectoryName(location);
+
+            return FindLibraryOrThrow(basepath, relativePath);
+        }
+
+        private SupportedPlatform GetCurrentPlatform()
+        {
+#if NETSTANDARD1_5
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                return SupportedPlatform.MacOS;
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                return SupportedPlatform.Linux;
+            }
+            else
+#endif
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return SupportedPlatform.Windows;
+            }
+
+            throw new InvalidOperationException("Current platform is not supported by LibraryLoader.");
+        }
+
+        private void ThrowIfNotSupportedPlatform()
+        {
```

Comment:
This method is still required

Reply:
I think this method should be called `ThrowIfNot64BitProcess`.

Reply:
Also... this might be more appropriately checked in the `LibraryLocator` in the Snappy files..

I don't think `LibraryLoader` *itself* requires a 64 bit process.

It might be the case that the 64 bit only restriction is a Snappy restriction?

Reply:
You are not mistaken... we decided to only support 64-bit Snappy.

My point was whether this check belonged in the general purpose LibraryLoader or somewhere Snappy specific like the new SnappyLocator class.

Either way is fine I guess.

Reply:
I think, in the current implementation, this logic should not be placed on the `Snappy` specific level since I don't see any cases when 32bits library can be used in the driver at all. So, no need to place this logic on lower levels(snappy specific), because the higher level(LibraryLoader) knows everything that he needs to decide the right conclusion regardless it's snappy or not. The answer will be depended only on what bitness of the current platform.